File: src/main/java/org/hisp/dhis/rules/models/RuleVariableCurrentEvent.java
package org.hisp.dhis.rules.models;

import org.hisp.dhis.rules.Option;
import org.hisp.dhis.rules.RuleVariableValue;
import org.hisp.dhis.rules.RuleVariableValueMapBuilder;

import javax.annotation.Nonnull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.hisp.dhis.rules.Utils.getLastUpdateDate;

public record RuleVariableCurrentEvent(
        @Nonnull String name,
        boolean useCodeForOptionSet,
        @Nonnull List<Option> options,
        @Nonnull String dataElement,
        @Nonnull RuleValueType dataElementType
) implements RuleVariableDataElement {

    @Nonnull
    public static RuleVariableCurrentEvent create(@Nonnull String name,
                                                  @Nonnull String dataElement, @Nonnull RuleValueType dataElementValueType, boolean useCodeForOptionSet, List<Option> options) {
        return new RuleVariableCurrentEvent(name, useCodeForOptionSet, options, dataElement, dataElementValueType);
    }

    @Override
    public Map<String, RuleVariableValue> createValues(RuleVariableValueMapBuilder builder,
                                                       Map<String, List<RuleDataValue>> allEventValues,
                                                       Map<String, RuleAttributeValue> currentEnrollmentValues,
                                                       Map<String, RuleDataValue> currentEventValues) {
        Map<String, RuleVariableValue> valueMap = new HashMap<>();

        RuleVariableValue variableValue;

        if (currentEventValues.containsKey(this.dataElement())) {
            RuleDataValue value = currentEventValues.get(this.dataElement());

            String optionValue = this.useCodeForOptionSet() ? value.value() : getOptionName(value.value());

            variableValue = RuleVariableValue.create(optionValue, this.dataElementType(),
                    List.of(optionValue), getLastUpdateDate(List.of(value)));
        } else {
            variableValue = RuleVariableValue.create(this.dataElementType());
        }

        valueMap.put(this.name(), variableValue);
        return valueMap;
    }
}

File: src/main/java/org/hisp/dhis/rules/models/RuleVariableCurrentEvent.kt
package org.hisp.dhis.rules.models

import org.hisp.dhis.rules.*
import javax.annotation.Nonnull

class RuleVariableCurrentEvent(
    val name: String,
    val useCodeForOptionSet: Boolean,
    val options: List<Option>,
    val dataElement: String,
    val dataElementType: RuleValueType
) : RuleVariableDataElement {
    override fun dataElement(): String {
        return dataElement
    }

    override fun dataElementType(): RuleValueType {
        return dataElementType
    }

    override fun options(): List<Option> {
        return options
    }

    override fun createValues(
        builder: RuleVariableValueMapBuilder,
        allEventValues: Map<String, List<RuleDataValue>>,
        currentEnrollmentValues: Map<String, RuleAttributeValue>,
        currentEventValues: Map<String, RuleDataValue>
    ): Map<String, RuleVariableValue> {
        val valueMap: MutableMap<String, RuleVariableValue> = HashMap()
        val variableValue: RuleVariableValue
        variableValue = if (currentEventValues.containsKey(dataElement)) {
            val value = currentEventValues[dataElement]
            val optionValue = if (useCodeForOptionSet) value!!.value else getOptionName(value!!.value)!!
            RuleVariableValue.create(
                optionValue, dataElementType,
                java.util.List.of(optionValue), getLastUpdateDate(java.util.List.of(value))
            )
        } else {
            RuleVariableValue.create(dataElementType)
        }
        valueMap[name] = variableValue
        return valueMap
    }
}
